interfaces/cython/cantera/reaction.pyx

```diff
@@ -24,32 +29,121 @@ cdef class _ReactionRate:
         will raise an exception.
         """
         if pressure:
-            self.base.update(temperature, pressure)
-            return self.base.eval(temperature, pressure)
+            self.rate.update(temperature, pressure)
+            return self.rate.eval(temperature, pressure)
         else:
-            self.base.update(temperature)
-            return self.base.eval(temperature)
+            self.rate.update(temperature)
+            return self.rate.eval(temperature)
+
+    @staticmethod
+    cdef wrap(shared_ptr[CxxReactionRateBase] rate):
+        """
+        Wrap a C++ Reaction object with a Python object of the correct derived type.
+        """
+        # ensure all reaction types are registered
+        if not(_reaction_rate_class_registry):
+            def register_subclasses(cls):
+                for c in cls.__subclasses__():
+                    rate_type = getattr(c, "_reaction_rate_type")
+                    _reaction_rate_class_registry[rate_type] = c
+                    register_subclasses(c)
+
+            # update global reaction class registry
+            register_subclasses(ReactionRate)
+
+        # identify class
+        rate_type = pystr(rate.get().type())
+        cls = _reaction_rate_class_registry.get(rate_type, ReactionRate)
+
+        # wrap C++ reaction rate
+        cdef ReactionRate rr
+        rr = cls(init=False)
+        rr._rate = rate
+        rr.rate = rr._rate.get()
+        return rr
+
+    @classmethod
+    def from_dict(cls, data, Kinetics kinetics=None):
+        """
+        Create a `ReactionRate` object from a dictionary corresponding to its YAML
+        representation.
+
+        An example for the creation of a ReactionRate from a dictionary is::
+
+            rate = ReactionRate.from_dict(
+                {"rate-constant": {"A": 38.7, "b": 2.7, "Ea": 26191840.0}},
+                kinetics=gas)
+
+        In the example, *gas* is a Kinetics (or Solution) object.
+
+        :param data:
+            A dictionary corresponding to the YAML representation.
+        :param kinetics:
+            A `Kinetics` object whose associated phase(s) contain the species
+            involved in the reaction.
+        """
+        if cls._reaction_rate_type != "":
+            raise TypeError(
+                "Class method 'from_dict' was invoked from '{}' but should "
+                "be called from base class 'Reaction'".format(cls.__name__))
```

Comment:
This is not related to Ray's comment, but it applies to the same places :smile: I think we should standardize on a formatting method for strings in new code (similar to how we prefer double quotes in general). I think we can use `f` strings for most cases, rather than `.format()`. They tend to be shorter and the value to be included is clearer, since it's placed directly in the braces.

The exception is that `f` strings can't be assigned to a variable and reused. In cases where you want to reuse a string, the `.format()` method would have to be used (or the built-in string interpolation in classes derived from `Exception`).

Reply:
@speth - absolutely, thanks!

@bryanwweber  ... hadn't used f-strings much until recently, but tend to agree that they're more succinct.